usr/lib/common/mech_openssl.c

```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
+    CK_BYTE *priv_exp = NULL, *buf = NULL;
+    EVP_PKEY *pkey = NULL;
+    EVP_MD_CTX *mdctx = NULL;
+    const EVP_MD *md = NULL;
+    size_t md_len;
+    unsigned char d_hash[SHA256_HASH_SIZE] = { 0 };
+    CK_RV rc;
+
+    /*
+     * The implementation of this function is copied from OpenSSL's function
+     * derive_kdk() in crypto/rsa/rsa_ossl.c and is slightly modified to fit to
+     * the OpenCryptoki environment.
+     * Changes include:
+     * - Different variable and define names.
+     * - Usage of TRACE_ERROR to report errors and issue debug messages.
+     * - Different return codes.
+     * - Different code to get the private key component 'd'.
+     * - Use of the EVP APIs instead of the internal APIs for Digest and HMAC
+     *   operations.
+     */
+
+    if (kdklen != SHA256_HASH_SIZE) {
+        TRACE_ERROR("KDK length is wrong\n");
+        return CKR_ARGUMENTS_BAD;
+    }
+
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_MODULUS,
+                                          &modulus);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_MODULUS\n");
+        return rc;
+    }
+
+    buf = calloc(1, modulus->ulValueLen);
+    if (buf == NULL) {
+        TRACE_ERROR("Failed to allocate a buffer for private exponent\n");
+        return CKR_HOST_MEMORY;
+    }
+
+    rc = template_attribute_get_non_empty(key_obj->template,
+                                          CKA_PRIVATE_EXPONENT, &priv_exp_attr);
+    if (rc != CKR_OK && rc != CKR_TEMPLATE_INCOMPLETE) {
+        TRACE_ERROR("Failed to get CKA_PRIVATE_EXPONENT\n");
+        goto out;
+    }
+
+    if (priv_exp_attr == NULL) {
+        rc = calc_rsa_priv_exp(tokdata, key_obj, buf, modulus->ulValueLen);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("calc_rsa_priv_exp failed\n");
+            goto out;
+        }
+        priv_exp = buf;
+    } else {
+        if (priv_exp_attr->ulValueLen < modulus->ulValueLen) {
+            memcpy(buf + modulus->ulValueLen - priv_exp_attr->ulValueLen,
+                   priv_exp_attr->pValue, priv_exp_attr->ulValueLen);
+            priv_exp = buf;
+        } else {
+            priv_exp = (CK_BYTE *)priv_exp_attr->pValue +
+                            priv_exp_attr->ulValueLen - modulus->ulValueLen;
+        }
+    }
+
+    /*
+     * we use hardcoded hash so that migrating between versions that use
+     * different hash doesn't provide a Bleichenbacher oracle:
+     * if the attacker can see that different versions return different
+     * messages for the same ciphertext, they'll know that the message is
+     * syntethically generated, which means that the padding check failed
+     */
+    md = EVP_sha256();
+    if (md == NULL) {
+        TRACE_ERROR("EVP_sha256 failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    if (EVP_Digest(priv_exp, modulus->ulValueLen, d_hash, NULL,
+                   md, NULL) <= 0) {
+        TRACE_ERROR("EVP_Digest failed\n");
```

Comment:
Just a general thing: at other places in opencryptoki, we print out openssl errors with something like ERR_print_errors_cb(openssl_err_cb, NULL); Would this help here also?

Reply:
Yes, in p11sak, we do print out the OpenSSL errors at the places where they happend.
But for the OCK libraries we don't. Instead, the API layer will print out all collected OpenSSL errors to the trace (it trace is enabled) in the END_OPENSSL_LIBCTX macro that ends every API function. 
So no need to do it here.